### PR TITLE
BUG: find_package(ITK ...) clobbered existing

### PR DESCRIPTION
**Pull request by hjmjohnson**
The find_package(ITK ...) with limited components caused
this second call to find_package(ITK) within GenerateCLP to
clobber the find_package(ITK) that an end user may have
called previously.  This would cause problems because
the end-user application now would only have access to the
limited set of modules that GenerateCLP needed.

### CODE REVIEW

**Comment by jcfr** [comment]
Thanks for the patch :+1: 

@hjmjohnson  While the patch is functional, could you confirm my understanding of the situation. See below.

Looking at `ModuleDescriptionParserConfig.cmake` and `UseModuleDescriptionParser.cmake`, they do not contain call to `find_package(ITK ...)`. 

Then, is it reasonable to say that calling `find_package(ITK COMPONENTS A B)` after calling `find_package(ITK COMPONENTS A B C)` is sufficient to modify the `ITK_LIBRARIES`, `ITK_INCLUDE_DIRS` and also generate (most likely) an incorrect `itk<factory_type>IOFactoryRegisterManager.h` as documented [here](https://github.com/InsightSoftwareConsortium/ITK/blob/533a2cd0e3c6c8abff45662f50eed9ad74a9afad/CMake/UseITK.cmake#L154-L161) ?

I would like to understand use cases causing problems to improve the ITK build system and associated IO Factory registration mechanism.


To summarize, the following sequence (copied from `UseGenerateCLP.cmake`) seems to cause problem:

```
find_package(ModuleDescriptionParser REQUIRED)
include(${ModuleDescriptionParser_USE_FILE})

# with ModuleDescriptionParser_ITK_COMPONENTS set to "ITKCommon ITKIOXML ITKEXPAT ITKExpat"

set(${PROJECT_NAME}_ITK_COMPONENTS
  ${ModuleDescriptionParser_ITK_COMPONENTS}
  )
find_package(ITK 4.3 COMPONENTS ${${PROJECT_NAME}_ITK_COMPONENTS} REQUIRED)
# Allow other projects to specify a different list of COMPONENTS
set(ITK_FOUND FALSE)
```

**Comment by jcfr** [comment]
Cc: @thewtex

**Comment by hjmjohnson** [comment]
I made a small little test case:

ccmake \
     -DITK_DIR:PATH=~/src/lung/ext_packages-bld/ITK-bld/ \
     -DSlicerExecutionModel_DIR:PATH=~/src/lung/ext_packages-bld/SlicerExecutionModel-bld/ \
     -DBUILD_AS_FAILURE:BOOL=ON \
     
[SEM_BUILD_Issues.tar.gz](https://github.com/Slicer/SlicerExecutionModel/files/229996/SEM_BUILD_Issues.tar.gz)
../SEM_BUILD_Issues/ 

======================
cmake .
-- ITK_LIBRARIES AFTER FIND ITK: itkdouble-conversion;itksys;itkvnl_algo;itkvnl;itkv3p_netlib;ITKCommon;itkNetlibSlatec;ITKStatistics;ITKTransform;ITKLabelMap;ITKMesh;itkzlib;ITKMetaIO;ITKSpatialObjects;ITKPath;ITKQuadEdgeMesh;ITKIOImageBase;ITKOptimizers;ITKPolynomials;ITKBiasCorrection;ITKBioCell;ITKDICOMParser;ITKEXPAT;ITKIOXML;ITKIOSpatialObjects;ITKFEM;gdcmDICT;gdcmMSFF;ITKznz;ITKniftiio;ITKgiftiio;itkhdf5_cpp;itkhdf5;ITKIOBMP;ITKIOBioRad;ITKIOCSV;ITKIOGDCM;ITKIOIPL;ITKIOGE;ITKIOGIPL;ITKIOHDF5;itkjpeg;ITKIOJPEG;itktiff;ITKIOTIFF;ITKIOLSM;ITKIOMRC;ITKIOMesh;ITKIOMeta;ITKIONIFTI;ITKNrrdIO;ITKIONRRD;itkpng;ITKIOPNG;ITKIOSiemens;ITKIOStimulate;ITKIOTransformBase;ITKIOTransformHDF5;ITKIOTransformInsightLegacy;ITKIOTransformMatlab;ITKIOVTK;ITKKLMRegionGrowing;ITKOptimizersv4;itkopenjpeg;ITKVTK;ITKWatersheds;ITKReview;ITKVideoCore;ITKVideoIO;ITKVtkGlue
-- ITK_LIBRARIES AFTER FIND SEM: ITKCommon;itksys;itkvnl_algo;itkvnl;itkv3p_netlib;ITKIOXML;ITKIOImageBase;ITKEXPAT
-- Configuring done
-- Generating done
-- Build files have been written to: /scratch/johnsonhj/xx-blc

**Comment by jcfr** [comment]
Thanks for the update.

I will add a test to SEM and integrate your patch.

Medium term, I also have an idea to improve ITK and address the problem generally.

**Comment by hjmjohnson** [comment]
I am very interested to have a solution for ITK proper.  This is a really nasty bug that crops up for both ITK and VTK, and SEM.  For end-users it can be really frustrating to try to figure these problems out.

Even if there was a way to throw some warning message, it would be great to put into place.

**Comment by jcfr** [comment]
Here is a proposal to help address the issue: https://docs.google.com/document/d/1n67HYbcjnQkmr5DZOm7gL1wPP1L88nGlONhLntWBLoE/edit

**Comment by jcfr** [comment]
Merging changes. We now have a path forward document in the google doc linked above. The SlicerExecutionModel will be updated once ITK has been updated.
